File: src/test/console/python/tests/androidAppTest.py
import unittest
import re
import time
from appium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.remote.remote_connection import RemoteConnection
import sys
sys.path.append('../..')
from python import configs
from python.utils.AutomationUtils import *
from python.services.DeviceService import DeviceService

class AndroidAppTest(unittest.TestCase):

  def setUp(self):
    self._command_executor = RemoteConnection(kobitonServerUrl())
    self._command_executor.set_timeout(configs.session_timeout)
    device = DeviceService().getOnlineDevice('Android')
    print('setUp - device:', device)
    self.driver = webdriver.Remote(self._command_executor, desiredCapabilitiesAndroidApp(device))

  def tearDown(self):
    self.driver.quit()

  def test_android_app(self):
    self.search_IFixit_on_home_screen()

  def search_IFixit_on_home_screen(self):
    print ('should allow to search iFixit on Home screen')
    self.driver.launch_app()
    time.sleep(2)
    self.driver.find_element_by_xpath("//*[@text='Search']").click()
    self.driver.find_element_by_xpath("//*[@resource-id='com.dozuki.ifixit:id/abs__search_src_text']").send_keys('Macbook Pro 2015')
    time.sleep(2)
    self.driver.press_keycode(66)
    time.sleep(5)
    firstResult = self.driver.find_element_by_xpath("//*[@resource-id='com.dozuki.ifixit:id/search_result_count']").text
    self.driver.find_element_by_xpath("//*[@resource-id='android:id/text1' and @text='Guides']").click()
    self.driver.find_element_by_xpath("//*[@resource-id='android:id/text1' and @text='Devices']").click()
    time.sleep(2)

    firstResult = re.findall('\d+', firstResult)[0]
    self.assertTrue(int(firstResult) >= 47,
      'The expected results are greater or equal to 47 results.')

  def swipe(self, element, direction = 'RightLeft'):
    # Get the size of screen.
    size = self.driver.get_window_size()

    startx = size['width'] * 0.90
    # Find endx point which is at left side of screen.
    endx = size['width'] * 0.10
    # Find vertical point where you wants to swipe. It is in middle of screen height.
    starty = size['height'] * 0.50

    if direction == 'RightLeft':
      # Swipe from Right to Left.
      self.driver.swipe(startx, starty, endx, starty, 200)

    if direction == 'LeftRight':
      # Swipe from Left to Right.
      self.driver.swipe(endx, starty, startx, starty, 200)

    time.sleep(5)

if __name__ == '__main__':
  suite = unittest.TestLoader().loadTestsFromTestCase(AndroidAppTest)
  isSuccess = unittest.TextTestRunner(verbosity=2).run(suite).wasSuccessful()
  sys.exit(not isSuccess)

File: java/java_testng_junit/src/main/java/kobiton/com/testng/AndroidAppTest.java
package kobiton.com.testng;

import java.util.concurrent.TimeUnit;
import java.util.HashMap;
import java.util.Map;

import com.google.common.collect.Lists;
import org.openqa.selenium.By;
import org.openqa.selenium.Dimension;
import org.openqa.selenium.Keys;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.WebDriverWait;
import org.testng.Assert;
import org.testng.annotations.AfterTest;
import org.testng.annotations.BeforeTest;
import org.testng.annotations.Test;
import io.appium.java_client.android.AndroidDriver;
import kobiton.com.configs.Configs;

public class AndroidAppTest {
	public static AndroidDriver<WebElement> driver = null;

	@BeforeTest
	public void Setup() {
		driver = new AndroidDriver<WebElement>(Configs.kobitonServerUrl(), Configs.desiredCapabilitiesAndroidApp());
		driver.manage().timeouts().implicitlyWait(120, TimeUnit.SECONDS);
	}

	@AfterTest
	public void Teardown() {
		try {
			if (driver != null)
				driver.quit();
		} catch (Exception e) {
			e.printStackTrace();
		}
	}

	@Test(priority = 1, description = "should allow to search some questions on Acura Support Community")
	public void testSearchQuestionsOnAcuraSupportCommunity() {

		/*
		 * Steps: 
		 * 1. Click on "Car and Truck" Categories on Homepage 
		 * 2. Click on "Acura" Categories
		 * 
		 * Expected: 
		 * 1. General Information is "Acura".
		 * 2.Verify five devices below displays.
		 * + Acura Integra
		 * + Acura MDX
		 * + Acura RL
		 * + Acura TL
		 * + Acura TSX
		 */

		driver.findElementByXPath("//*[@resource-id='android:id/home']").click();
		sleep(2);
		driver.findElementByXPath("//*[@text='Car and Truck']").click();
		driver.findElementByXPath("//*[@text='Acura']").click();
		sleep(2);
		String acuraText = driver
			.findElementByXPath("//*[@resource-id='com.dozuki.ifixit:id/topic_title' and @index=1]").getText();
		String acuraIntegraText = driver
			.findElementByXPath("//*[@resource-id='com.dozuki.ifixit:id/topic_title' and @index=3]").getText();
		String acuraMDXText = driver
			.findElementByXPath("//*[@resource-id='com.dozuki.ifixit:id/topic_title' and @index=4]").getText();
		String acuraRLText = driver
			.findElementByXPath("//*[@resource-id='com.dozuki.ifixit:id/topic_title' and @index=5]").getText();
		String acuraTLText = driver
			.findElementByXPath("//*[@resource-id='com.dozuki.ifixit:id/topic_title' and @index=6]").getText();
		String acuraTSXText = driver
			.findElementByXPath("//*[@resource-id='com.dozuki.ifixit:id/topic_title' and @index=7]").getText();

		Assert.assertEquals(acuraText, "Acura");
		Assert.assertEquals(acuraIntegraText, "Acura Integra");
		Assert.assertEquals(acuraMDXText, "Acura MDX");
		Assert.assertEquals(acuraRLText, "Acura RL");
		Assert.assertEquals(acuraTLText, "Acura TL");
		Assert.assertEquals(acuraTSXText, "Acura TSX");
	}

	@Test(priority = 2, description = "should allow to search iFixit on Home screen")
	public void testSearchIFixitOnHomeScreen() {

		/*
		 * Steps: 
		 * 1. Reopen iFixit app 
		 * 2. Click Search menu on the left menu bar
		 * 3. Search keyword 'Macbook Pro 2015' 
		 * 4. Press Enter button on keyboard
		 * 
		 * Expected: It should show at least 48 results.
		 * 
		 * 5. Select Devices item on the Guides/Devices dropdown
		 * 
		 * Expected: It should show at least 5 results.
		 */

		driver.launchApp();
		sleep(2);
		driver.findElementByXPath("//*[@resource-id='com.dozuki.ifixit:id/action_search']").click();
		driver.findElementByXPath("//*[@resource-id='com.dozuki.ifixit:id/abs__search_src_text']")
			.sendKeys("Macbook Pro 2015");
		driver.pressKeyCode(66);
		sleep(2);
		String firstResult = driver.findElementByXPath("//*[@resource-id='com.dozuki.ifixit:id/search_result_count']")
			.getText();
		driver.findElementByXPath("//*[@resource-id='android:id/text1' and @text='Guides']").click();
		driver.findElementByXPath("//*[@resource-id='android:id/text1' and @text='Devices']").click();

		String secondResult = driver.findElementByXPath("//*[@resource-id='com.dozuki.ifixit:id/search_result_count']")
			.getText();
		
		firstResult = firstResult.replaceAll("[^0-9]", "");
		Assert.assertTrue(Integer.parseInt(firstResult) >= 48,
			"The returned results are greater or equal to 48 results.");
		secondResult = secondResult.replaceAll("[^0-9]", "");
		Assert.assertTrue(Integer.parseInt(secondResult) >= 5,
			"The returned results are greater or equal to 5 results.");
	}

	@Test(priority = 3, description = "should allow to send command mobile shell")
	public void testMobileShell() {

		/*
		 * Steps:
		 * 1. Send mobile:shell command line to device
		 *
		 * Expected: It should allow to send command mobile shell.
		 */

		Map<String, Object> args = new HashMap<String, Object>();
		args.put("command", "echo");
		args.put("args", Lists.newArrayList("hello", "world"));

		Object output = driver.executeScript("mobile: shell", args);
		Assert.assertEquals(output, "hello world");
	}

	public void sleep(int seconds) {
		try {
			Thread.sleep(seconds * 1000);
		} catch (InterruptedException e) {
			e.printStackTrace();
		}
	}
}
